### ISSUE TITLE
Feature Request: Plugins for effects

### ISSUE THREAD

**Issue by teejayhh** [nolimits4web/Swiper#1568]
Hello, I am not sure if this is the right place but I thought I post it here. Today I came across a problem with the fade effect were I wanted a different behaviour of fading slides. Currently there are two options of fading, fade slides on top of each other or crossfade slides.
This works all really well when you fade solid elements like images on top of each other. As soon as you deal with text slides fading becomes very ugly. 
What I wanted to do is that the current slide fades away first and once this is done the new slide fades in, animation should not overlap. 

I can sort of create this behaviour by setting this 

``` css
.swiper-slide-active {
    transition-delay: 1s;
}
```

which works great when you interact with the pageinator or arrow navigation, but as soon as touch events and progress calculation comes in place the css properties are directly applied on the dom elements and from that point they are applied on all affected slides at the same time.

it would be cool to add a mechanism to add effects as plugins which need to implement an interface , like defaults, setTransition,setTranslate  and so on. It kind of feels like the right thing todo, but that might be just me.

Cheers Thomas

**Comment by nolimits4web** [comment]
Hi, those `setTransition` `setTranslate` effect methods are basically just callbacks, so you can just take the code of fade effect and put same logic with requred changes to the `onSetTransition` and `onSetTranslate` callbacks

**Comment by teejayhh** [comment]
I know about the callbacks. I just thought it would be cleaner to offload the code in its own classes. That way it would be simple to create a repo of new effects which can be loaded on demand.

**Comment by nolimits4web** [comment]
Got it. It is possible via Plugins API (which is not yet documented):

my-effect.js
```js
Swiper.prototype.plugins.myEffect = function (swiper) {
    if (swiper.params.effect !== 'my-effect') return;
    // Do something on plugin init
    // ...

    // Return plugin hooks (same as callbacks with same arguments)
    return {
        onSetTranslate: function () {
            // Do something
        },
        onSetTransition: function () {
            // Do something
        }
    };
};
```

**Comment by markleung1969** [comment]
it solves my problem !  thx ! !

**Comment by anteksiler** [comment]
Hi @nolimits4web,

I am trying to make a vertical 3d circle using your code:
`
Swiper.prototype.plugins.circle3d = function (swiper) {
				  if (swiper.params.effect !== 'circle3d') return;
				
				    // Return plugin hooks (same as callbacks with same arguments)
					return {
					  onSetTranslate: function () {
					    var wrapperRotate = 0;
					    for (var i = 0; i < swiper.slides.length; i++) {
					        var slide = swiper.slides.eq(i);
					        var slideAngle = i * 60;
					        var round = Math.floor(slideAngle / 360);
					        var transform = 'rotateX(' + slideAngle + 'deg) translateZ('+(swiper.size / 2) +'px)';
									var progress = Math.max(Math.min(swiper.slides.eq(0).progress, 1), -1);
									if (progress <= 1 && progress > -1) {
									    wrapperRotate = i * 60;
									}
									
					        slide.transform(transform);
					    }
							swiper.wrapper.css({
							    '-webkit-transform-origin': '50% 50% -' + (swiper.size / 2) + 'px',
							    '-moz-transform-origin': '50% 50% -' + (swiper.size / 2) + 'px',
							    '-ms-transform-origin': '50% 50% -' + (swiper.size / 2) + 'px',
							    'transform-origin': '50% 50% -' + (swiper.size / 2) + 'px'
							});
              swiper.wrapper.transform('rotateX(' + wrapperRotate + 'deg)');
						}
					};
				};
`

I am trying to achieve:

![image](https://cloud.githubusercontent.com/assets/1000992/17230485/0733bdb4-5525-11e6-93a4-ea4a7c8d8099.png)

I am stuck, so any help is appreciated.

**Comment by nolimits4web** [comment]
Issue is closed because of outdated/irrelevant/not actual/needed more information/inactivity.

If this issue is still actual and reproducible for latest version of Swiper, please create new issue and fill the issue template correctly:
* Clearly describe the issue including steps to reproduce when it is a bug.
* Make sure you fill in the earliest version that you know has the issue.
* Provide live link or JSFiddle/Codepen or website with issue
